### PR TITLE
parking_lot Mutex/Condvar via feature gate (take 2)

### PR DESCRIPTION
**Pull request by dekellum**
This is a second attempt at #2130, as discussed on discord.

## Motivation

* The _parking_lot_ crate's `Mutex` and `Condvar` still best _libstd_ in benchmarks.  
* From Amanieu/parking_lot#119 (and linked rust-lang PR), it does not appear (AFAICT) that replacing libstd's current implementations with _parking_lot_ inspired types is particularly close to being accepted. (When that does happen and in MSRV, this PR should be easy enough to revert.)

## Solution

With a non-default, not included in "full" *parking_lot* feature enabled, new `Mutex` and `Condvar` adapter types are included which bridge the parking_lot types to the std::sync interface.  Note that for compatibility, the adapters return types in the form `Result<_, *LockError>` but never actually error, because _parking_lot_ does not have poisoning. To quote https://docs.rs/parking_lot/0.10.0/parking_lot/type.Mutex.html:

> * No poisoning, the lock is released normally on panic.

I have not found a case in the current codebase where tokio attempts to recover from a poisoned `Mutex`, or is dependent `panic!` from posion error `unwrap`-ing.

## Possible Followup Steps

IMO this should be deferred until after reviewing and hopefully merging this PR:

1. Not all references to `Mutex` and `Condvar` in the tree go through the `loom` module. It may or may not make sense to replace all of these cases, including in tests? 

``` txt
git grep -n -E  "use.*Mutex" | grep -v "loom::" | grep -v "tokio::sync" | grep -v "src/loom"
tokio-test/src/task.rs:9:use std::sync::{Arc, Condvar, Mutex};
tokio/src/process/unix/orphan.rs:3:use std::sync::Mutex;
tokio/src/runtime/tests/loom_oneshot.rs:3:use std::sync::{Arc, Mutex};
tokio/src/signal/registry.rs:11:use std::sync::Mutex;
tokio/src/sync/barrier.rs:3:use std::sync::Mutex;
tokio/src/sync/mod.rs:27:    pub use mutex::{Mutex, MutexGuard};
tokio/src/sync/watch.rs:61:use std::sync::{Arc, Mutex, RwLock, RwLockReadGuard, Weak};
tokio/src/task/error.rs:4:use std::sync::Mutex;
tokio/src/tests/loom_schedule.rs:5:use std::sync::Mutex;
tokio/src/tests/mock_schedule.rs:5:use std::sync::Mutex;
tokio/src/time/clock.rs:38:    use std::sync::{Arc, Mutex};
tokio/tests/fs_dir.rs:7:use std::sync::{Arc, Mutex};
tokio/tests/io_driver.rs:12:use std::sync::{mpsc, Arc, Mutex};
tokio/tests/support/mock_file.rs:10:use std::sync::{Arc, Mutex};
```
## Benchmarks

See results comparing parking_lot enabled vs disabled in #2130. I did not find any significant difference in results with the new implementation in this PR.

### CODE REVIEW

File: tokio/Cargo.toml [hunk 1]
```diff
@@ -105,8 +104,8 @@ memchr = { version = "2.2", optional = true }
 mio = { version = "0.6.20", optional = true }
 iovec = { version = "0.1.4", optional = true }
 num_cpus = { version = "1.8.0", optional = true }
-# Backs `DelayQueue`
-slab = { version = "0.4.1", optional = true }
+parking_lot = { version = "0.10.0", optional = true }
```
Comment (carllerche) [inline tokio/Cargo.toml]: We probably should add a comment saying that, including the `parking_lot` feature flag will mean adopting their MSRV as well.

Comment (dekellum) [reply tokio/Cargo.toml]: Some (all?) other features are discussed in src/lib.rs rustdoc. Would that be a good place for this? Seems like the Cargo.toml would make such a bit hard to find?

Comment (carllerche) [reply tokio/Cargo.toml]: Yes, we can add it in `lib.rs` as well.

Maybe we should just say "if you enable this feature, you get no MSRV guarantee" :)

Comment (dekellum) [reply tokio/Cargo.toml]: In bbb4356 I add `parking_lot` to src/lib documented features. It forced a clarification that `Full` is not "all features" but rather "all public API features".  Also I tried to strike a direct, fact-based tone by saying "MSRV may increase according to the _parking_lot_ release in use.".  As this is getting into Tokio project policy territory, I feel you might want to tweak this further to your liking post merge?

File: tokio/src/lib.rs [hunk 2]
```diff
@@ -66,13 +66,13 @@
 //! is possible to just enable certain features over others. By default, Tokio
 //! does not enable any features but allows one to enable a subset for their use
 //! case. Below is a list of the available feature flags. You may also notice
-//! above each function, struct and trait there is a set of feature flags
-//! that are required for that item to be enabled. If you are new to Tokio it is
-//! recommended that you use the `full` feature flag which will enable everything.
+//! above each function, struct and trait there is listed one or more feature flags
+//! that are required for that item to be used. If you are new to Tokio it is
+//! recommended that you use the `full` feature flag which will enable all public APIs.
```
Comment (jonhoo) [inline tokio/src/lib.rs]: Some of these changes seem unrelated?

Comment (dekellum) [reply tokio/src/lib.rs]: The new `parking_lot` feature is not included in `full` so I needed to distinguish public API enabling features from internal features (this being the first).  So saying full enables "everything" doesn't work so well anymore, could be confusing.

**Comment by jonhoo** [comment]
As for the question around `loom` types, yes, I think we should make sure everything goes through `loom`.

**Comment by dekellum** [comment]
> Is thare any code that does `.lock().unwrap()? Presumably we are changing the behavior of that here, no?

Yes the whole code base does `lock.unwrap()` so your concern over changing behavior of cascading panics is correct.  Keep in mind this is opt-in only. Its also not entirely clear to me which behavior will be better... I guess it depends on individual circumstance? Personally I would prefer if tokio has configurations on all thread pools (reactor and spawn blocking pool) to just abort-on-panic.
